### PR TITLE
Remove TrainingTypePlugin.on_save and Accelerator.on_save

### LINKED ISSUE #8987: Deprecate TrainingTypePlugin.on_save
## 🚀 Feature

Deprecate this method off the TrainingTypePlugin: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/6de66eb110a63d67dc2ebb74e149981cd93aa431/pytorch_lightning/plugins/training_type/training_type_plugin.py#L204-L206

cc @four4fish

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

This method is redundant as we also have `save_checkpoint` inside the training type plugin as well: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/6de66eb110a63d67dc2ebb74e149981cd93aa431/pytorch_lightning/plugins/training_type/training_type_plugin.py#L269-L279

Therefore, the plugin already has the ability to augment/modify the checkpoint dict as needed, and `on_save` doesn't need to exist on the public API. it is an implementation detail of individual plugins

### Pitch

- Deprecate this method in v1.5
- Remove this method in v1.7

@awaelchli @tchaton @SeanNaren @justusschock as the plugin API is still in beta, does the 2 minor version cadence for deprecations still apply? 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?

Fixes #8987 
Double checked none of the plugins overridden on_save()

### Does your PR introduce any breaking changes? If yes, please list them.

Removes `on_save` from `TrainingTypePlugin` and `Accelerator`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
thanks for picking this up @four4fish !

@PyTorchLightning/core-contributors I'm not sure what the deprecation policies & versioning guarantees are for plugins & accelerators.

the docs say this:

> The Plugin API is in beta and subject to change. For help setting up custom plugins/accelerators, please reach out to us at support@pytorchlightning.ai

https://pytorch-lightning.readthedocs.io/en/latest/extensions/plugins.html

Does this mean we can make the change in-place like this? 

deprecation policy aside, this PR is what we'll end up doing.

File: CHANGELOG.md [hunk 1]
```diff
@@ -111,7 +111,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `DataModule` properties: `train_transforms`, `val_transforms`, `test_transforms`, `size`, `dims` ([#8851](https://github.com/PyTorchLightning/pytorch-lightning/pull/8851))
 
 
--
+- Deprecated `TrainingTypePlugin.on_save` and `Accelerator.on_save` ([#8987](https://github.com/PyTorchLightning/pytorch-lightning/pull/8987))
```
Comment (ananthsub) [inline CHANGELOG.md:114]: this would be a removal as we're deleting the functionality altogether. the typical procedure is to mark the API as deprecated, and then make a breaking change a few versions later. if this approach is okay, we should move this line to the Removed section below

Comment (four4fish) [reply CHANGELOG.md:114]: Thank you for pointing it out @ananthsub ! What's your recommendation for the correct approach? 
I feel the save_checkpoint() is already a replacement for on_save(). The on_save() function just return the param it got, so I think it's safe to remove it directly. What's your advise on this? Should we always deprecate first then remove in next few versions?

Comment (ananthsub) [reply CHANGELOG.md:114]: I agree. Let's add this directly to the `Removed` section

**Review by ananthsub — APPROVED** [review]
Discussing with @carmocca - the plugins & accelerator API is not stable enough yet for it to warrant full backward compatibility. the blast radius of this change is small given none of the framework's implementations are affected. Therefore, I'm in favor of directly removing this.

**Review by carmocca — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -157,6 +157,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated `GradInformation` module in favor of `pytorch_lightning.utilities.grads` ([#8831](https://github.com/PyTorchLightning/pytorch-lightning/pull/8831/))
 
 
+- Removed `TrainingTypePlugin.on_save` and `Accelerator.on_save` ([#8987](https://github.com/PyTorchLightning/pytorch-lightning/pull/8987))
```
Comment (awaelchli) [inline CHANGELOG.md:160]:
```suggestion
- Removed `TrainingTypePlugin.on_save` and `Accelerator.on_save` ([#9023](https://github.com/PyTorchLightning/pytorch-lightning/pull/9023))
```

**Review by awaelchli — APPROVED** [review]
